SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,441 @@
+
+    return [mutableDeletes copy];
+}
+
+- (NSArray<SDLDeleteCommand *> *)sdl_deleteCommandsForVoiceCommands:(NSArray<SDLVoiceCommand *> *)voiceCommands {
+    NSMutableArray<SDLDeleteCommand *> *mutableDeletes = [NSMutableArray array];
+    for (SDLVoiceCommand *command in self.voiceCommands) {
+        SDLDeleteCommand *delete = [[SDLDeleteCommand alloc] initWithId:command.commandId];
+        [mutableDeletes addObject:delete];
+    }
+
+    return [mutableDeletes copy];
+}
+
+#pragma mark Commands / SubMenu RPCs
+
+- (NSArray<SDLRPCRequest *> *)sdl_mainMenuCommandsForCells:(NSArray<SDLMenuCell *> *)cells withArtwork:(BOOL)shouldHaveArtwork {
+    NSMutableArray<SDLRPCRequest *> *mutableCommands = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.subCells.count > 0) {
+            [mutableCommands addObject:[self sdl_subMenuCommandForMenuCell:cell]];
+        } else {
+            [mutableCommands addObject:[self sdl_commandForMenuCell:cell withArtwork:shouldHaveArtwork]];
+        }
+    }
+
+    return [mutableCommands copy];
+}
+
+- (NSArray<SDLRPCRequest *> *)sdl_subMenuCommandsForCells:(NSArray<SDLMenuCell *> *)cells withArtwork:(BOOL)shouldHaveArtwork {
+    NSMutableArray<SDLRPCRequest *> *mutableCommands = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.subCells.count > 0) {
+            [mutableCommands addObjectsFromArray:[self sdl_allCommandsForCells:cell.subCells withArtwork:shouldHaveArtwork]];
+        }
+    }
+
+    return [mutableCommands copy];
+}
+
+- (NSArray<SDLRPCRequest *> *)sdl_allCommandsForCells:(NSArray<SDLMenuCell *> *)cells withArtwork:(BOOL)shouldHaveArtwork {
+    NSMutableArray<SDLRPCRequest *> *mutableCommands = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.subCells.count > 0) {
+            [mutableCommands addObject:[self sdl_subMenuCommandForMenuCell:cell]];
+            [mutableCommands addObjectsFromArray:[self sdl_allCommandsForCells:cell.subCells withArtwork:shouldHaveArtwork]];
+        } else {
+            [mutableCommands addObject:[self sdl_commandForMenuCell:cell withArtwork:shouldHaveArtwork]];
+        }
+    }
+
+    return [mutableCommands copy];
+}
+
+- (SDLAddCommand *)sdl_commandForMenuCell:(SDLMenuCell *)cell withArtwork:(BOOL)shouldHaveArtwork {
+    SDLAddCommand *command = [[SDLAddCommand alloc] init];
+
+    SDLMenuParams *params = [[SDLMenuParams alloc] init];
+    params.menuName = cell.title;
+    params.parentID = cell.parentCellId != UINT32_MAX ? @(cell.parentCellId) : nil;
+
+    command.menuParams = params;
+    command.vrCommands = cell.voiceCommands;
+    command.cmdIcon = (cell.icon && shouldHaveArtwork) ? [[SDLImage alloc] initWithName:cell.icon.name] : nil;
+    command.cmdID = @(cell.cellId);
+
+    return command;
+}
+
+- (SDLAddSubMenu *)sdl_subMenuCommandForMenuCell:(SDLMenuCell *)cell {
+    return [[SDLAddSubMenu alloc] initWithId:cell.cellId menuName:cell.title];
```

Comment:
There should be an option to set the position of the submenu cells. Order is important for alphabetical lists.
![submenu items out of order on manticore](https://user-images.githubusercontent.com/19915981/38994165-8f450216-43b3-11e8-8164-e4298f813c3d.png)

Reply:
The order is set by how they're ordered in the array. The requests are sent in order, but not waiting for the previous response. I wonder if something with TCP is sending them out of order. I didn't want to implement `position` if I didn't have to, but maybe I'll have to.